content/300-guides/300-upgrade-guides/200-upgrading-versions/030-upgrading-to-prisma-4.mdx

```diff
@@ -0,0 +1,398 @@
+
+If you use Postgres, as a result of tightened type coercion rules you might need to update certain queries that worked in 3.x.
+
+#### Upgrade path
+
+We recommend that you re-test your usage of `$queryRaw` and double-check that the types you pass into your raw queries match the types that PostgreSQL expects.
+
+For example, in v4.0.0, the following query fails:
+
+```js
+await prisma.$queryRaw`select length(${42});
+```
+
+This is because PostgreSQL’s `length(string)` function expects a string as the parameter. To fix this, explicitly coerce the incoming number into a string as follows:
+
+```js
+await prisma.$queryRaw`select length(${42}::text);`
+```
+
+### <inlineCode>DbNull</inlineCode>, <inlineCode>JsonNull</inlineCode> and <inlineCode>AnyNull</inlineCode> are now objects
+
+JavaScript `null` is ambiguous for JSON columns, so Prisma uses `DbNull`, `JsonNull`, and `AnyNull` to distinguish between the database `NULL` value and the JSON `null` value. Before v4.0.0, `DbNull`, `JsonNull`, and `AnyNull` were string constants. From v4.0.0, they are objects.
+
+See [Filtering by null values](/concepts/components/prisma-client/working-with-fields/working-with-json-fields#filtering-by-null-values) for more information.
+
+#### Upgrade path
+
+If you use literal strings to address these values, then you must update your code to use the following named constants instead:
+
+- For `DbNull` use `Prisma.DbNull`
+- For `JsonNull` use `Prisma.JsonNull`
+- For `AnyNull`: use `Prisma.AnyNull`
+
+If you already use these named constants, then you do not need to take any action.
+
+If you now get a type error when you pass `Prisma.DbNull` as the value of a JSON field, then this probably indicates a bug in your code that our types did not catch before v4.0.0. The field where you tried to store `DbNull` is probably not nullable in your schema. As a result, a literal "DbNull" string was stored in the database instead of NULL.
+
+### Default composite types in MongoDB
+
+From v4.0.0, if you carry out a database read on a composite type when all of the following conditions are true, then Prisma Client inserts the default value into the result.
+
+Conditions:
+
+- The composite type is required, and
+- The composite type has a default value, and
+- The composite type is not present in the returned document or documents.
+
+This behavior is now consistent with the behavior for model fields.
+
+To learn more, see [Default values on required composite types](/concepts/components/prisma-client/composite-types#default-values-on-required-composite-types).
+
+#### Upgrade path
+
+If you currently rely on a return value of `undefined`, then you need to refactor your code to handle the default value that is now returned in Prisma 4.
```

Comment:
I'm pretty sure the QE doesn't support the `undefined` value and returns `null` instead.

```suggestion
If you currently rely on a return value of `null`, then you need to refactor your code to handle the default value that is now returned in Prisma 4.
```